File: utils/shuffle/shuffle.go
package shuffle

import (
	"strings"

	"github.com/navidrome/navidrome/utils/random"
)

// Keys are the attributes used to keep same-artist (and same-album) tracks
// from sitting next to each other after a shuffle.
type Keys struct {
	Artist string
	Album  string
}

// TrackKeys maps media-file artist/album fields onto shuffle Keys.
// Artist falls back to albumArtist when the track artist is empty.
func TrackKeys(artist, albumArtist, album string) Keys {
	if artist == "" {
		artist = albumArtist
	}
	return Keys{Artist: artist, Album: album}
}

// IsRandomSort reports whether a query sort is a random ordering
// (Native API `_sort=random`, SQL `random()`, or the SEEDEDRAND rewrite).
func IsRandomSort(sort string) bool {
	s := strings.ToLower(strings.TrimSpace(sort))
	return s == "random" || s == "random()" || strings.HasPrefix(s, "seededrand(")
}

// Slice is a full permutation of items: Fisher-Yates, then a light
// adjacent-repair pass. It never drops or duplicates items.
//
// Important: do NOT rebuild the list by always draining the largest artist
// pile — that made libraries dominated by a few artists (e.g. Unknown /
// Arctic Monkeys / Muse) play those artists almost exclusively.
func Slice[T any](items []T, keys func(T) Keys) {
	n := len(items)
	if n < 2 {
		return
	}

	for i := n - 1; i > 0; i-- {
		j := int(random.Int64N(i + 1))
		items[i], items[j] = items[j], items[i]
	}

	repairAdjacent(items, keys)
}

// repairAdjacent keeps the Fisher-Yates order and only swaps forward when
// two neighbors share an artist (then album) and a later break is available.
func repairAdjacent[T any](items []T, keys func(T) Keys) {
	n := len(items)
	for i := 1; i < n; i++ {
		prev := keys(items[i-1])
		cur := keys(items[i])

		if prev.Artist != "" && cur.Artist == prev.Artist {
			for j := i + 1; j < n; j++ {
				if keys(items[j]).Artist != prev.Artist {
					items[i], items[j] = items[j], items[i]
					cur = keys(items[i])
					break
				}
			}
		}

		if prev.Artist != "" && cur.Artist == prev.Artist &&
			prev.Album != "" && cur.Album == prev.Album {
			for j := i + 1; j < n; j++ {
				k := keys(items[j])
				if k.Artist != prev.Artist || k.Album != prev.Album {
					items[i], items[j] = items[j], items[i]
					break
				}
			}
		}
	}
}

File: ui/src/utils/shuffle.js
// Balanced shuffle: Fisher-Yates, then light adjacent artist/album repair.
// Full permutation (no drops/dupes). Does NOT drain largest artist piles —
// that over-played dominant artists (Unknown / Muse / etc).

const randomFloat = () => {
  if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
    const buf = new Uint32Array(1)
    crypto.getRandomValues(buf)
    return buf[0] / 2 ** 32
  }
  return Math.random()
}

const fisherYates = (items, random) => {
  for (let i = items.length - 1; i > 0; i--) {
    const j = Math.floor(random() * (i + 1))
    ;[items[i], items[j]] = [items[j], items[i]]
  }
}

export const songKeys = (song) => ({
  artist: song?.artist || song?.singer || song?.albumArtist || '',
  album: song?.album || '',
})

const repairAdjacent = (items, getKeys) => {
  for (let i = 1; i < items.length; i++) {
    const prev = getKeys(items[i - 1])
    let cur = getKeys(items[i])

    if (prev.artist && cur.artist === prev.artist) {
      for (let j = i + 1; j < items.length; j++) {
        if (getKeys(items[j]).artist !== prev.artist) {
          ;[items[i], items[j]] = [items[j], items[i]]
          cur = getKeys(items[i])
          break
        }
      }
    }

    if (
      prev.artist &&
      cur.artist === prev.artist &&
      prev.album &&
      cur.album === prev.album
    ) {
      for (let j = i + 1; j < items.length; j++) {
        const k = getKeys(items[j])
        if (k.artist !== prev.artist || k.album !== prev.album) {
          ;[items[i], items[j]] = [items[j], items[i]]
          break
        }
      }
    }
  }
  return items
}

export const balancedShuffle = (
  items,
  getKeys = songKeys,
  random = randomFloat,
) => {
  const arr = items.slice()
  if (arr.length < 2) {
    return arr
  }
  fisherYates(arr, random)
  return repairAdjacent(arr, getKeys)
}
